community/planning/ref_set.md

```diff
@@ -0,0 +1,436 @@
+
+## Motivation
+[motivation]: ##motivation
+
+This RFC intends to simplify and generalize the BlockManager for use with other
+objects, such as batches.  This will provide broader usage for library
+consumers.
+
+## Guide-level explanation
+[guide-level-explanation]: ##guide-level-explanation
+
+The `RefSet` provides references to data objects.  For example, it can be used
+to store blocks or batches.  The object manager's role is to ensure that only a
+single instance of an object exists in memory.
+
+The managed objects need to provide some value as an identifier. This ID will be
+used to reference the object by other components. A reference may be requested
+from the `RefSet` using a raw ID, but the manager will return a reference only
+if it contains the referenced object.
+
+The `RefSet` will provide access to its managed objects via leases.  These
+leases allow the caller to reference the object, but not modify it.
+Modifications may only be done by creating a new instance, with a new ID.
+
+Unlike the current `BlockManager`, the `RefSet` will rely on the underlying
+store to provide strong guarantees that an object with any references will not
+be deleted. Initial store implementations will rely on RDBSs to provide those
+guarantees, via foreign key constraints.
+
+Moving the referential integrity to the underlying storage system does mean that
+a component holding onto an object reference is not necessarily guaranteed that
+the object exists.  It does require that the component has performed a necessary
+upgrade of a raw ID before attempting to acquire a lease to the object.
+
+If the component holding a reference requires referential integrity, it must
+persist its own reference.  For example, all chain heads of interest must be
+persisted, to guarantee that blocks stored via a `RefSet<Block>` are preserved.
+
+In a SQL environment, this can be accomplished with two foreign key constraints.
+The first constraint is between a block ID and a block's `previous_block_id` in
+the primary block table.  The second constraint is between a `chain_head`
+record's block ID and a block in the primary block table. With these
+constraints, an individual block in the primary table cannot be deleted without
+the referencing record being deleted.
+
+References are held through the use of a `RefToken`.  These tokens convey
+ownership of a reference to an item in the `RefSet`.  That ownership may be
+transferred to another component.  In order to do so, the transferring component
+may provide a `Command` to be executed when the transfer is complete. The
+transferee may also provide a `Command` on acquisition of the token. These two
+commands will be executed within a transactional context, such that they are
+executed atomically.
+
+## Reference-level explanation
+[reference-level-explanation]: ##reference-level-explanation
+
+### Managed Objects
+
+Each object, `T`, with an identifier `ID`, stored in the has the following
```

Comment:
Should this say "stored in the RefSet has"

Reply:
Fixed